### DISCUSSION TITLE
Applying `egg` to SSA-based IRs

### DISCUSSION THREAD

**Discussion by femtomc** [egraphs-good/egg#145] [Q&A]
Hi! What are the prospects of utilizing `egg` to perform rewrite-driven transformations on SSA-based IRs (?)

I read this paper (whose authors appear to be collaborators with y'all): https://www.cs.cornell.edu/~ross/publications/eqsat/ where they describe `E-PEG`-based program representation.

On the other hand, I'm working on IRs which are closer to strict SSA (where all control flow is de-sugared to basic blocks and branching) and everything is linear (e.g. blocks are vectors of instructions) vs. strictly tree-like. 

I'd love to find or understand how equality saturation could be applied to this representation structure -- it's likely that maybe this information can be gleaned from the above paper (but I expect that maintainers here might be able to unpack this more than I can).

**Comment by chandrakananandi** [comment]
Hi! I use `egg` for a language which is mostly-SSA. In my case, things "just work" so far because I mostly care about straight line blocks. The only additional information I track which may be necessary for other SSA-based usages as well is to make sure you don't use variables before they are defined. This, in my case, was possible with a simple e-class analysis.

**Reply by femtomc** [reply]
I’m curious about this — are you working with single blocks at a time? How does scope work in your language — e.g. can blocks which dominate provide SSA references which do not need to be passed as block arguments (or otherwise represented in phi nodes)?

e.g. ignoring flow between blocks, I can sort of imagine just restricting attention to def-use chains, and then you have a nice DAG-like thing, and there’s probably some details about fixing the chains up after rewriting.

Re — is this project open source at all? I’m just getting started with egg so I’m curious about usage like this. No worries if not!

**Reply by chandrakananandi** [reply]
My use case at the moment involves working with a single block at a time which are all jump-free. I don't actually modify the code itself (it's part of a verification process, so I shouldn't :) ), but I generate a summary of linear invariants from the blocks. Your question about the dominating block is interesting --- I will need to run some more experiments to see what happens in that case. I will get back to you on this once I have more info.

Unfortunately the project is not open-source yet but I would like to make it available in some form (maybe an article or blog post) soon.

**Comment by mwillsey** [comment]
Hi, I think @chandrakananandi is right. I don't think there is any major technical blocker stopping you from doing this today!

The original eqsat paper you linked to contains a lot of good ideas on how to encode your problem into an e-graph. If I'm understanding your setting, you'll still want E-PEGs if you want to encode loops in a transparent way that you can optimize through. If you only care about optimizing one basic block (one DFG) at a time, then you don't need them.

The egg paper (and tool) are mostly innovating in how equality saturation is done. How you encode your problem is still up to you!

Despite there not being any huge blockers, I still think it's a large and challenging task, and one that I'd like to try to tackle at some point (if I can find the time), or see someone else take a stab at!

One thing that I will add: you'll have a much easier time with a tree- or dag-like IR that a linear, mutating IR. All the rewrites that you do are over trees or dags, and just overall the e-graph doesn't do a lot for you if your language is heavily sequential. If that's your case, consider building def-use chains or some other method to make things more graph-like for the e-graph.

**Reply by femtomc** [reply]
Are either of you familiar with MLIR? I’ve been working on a small scale Rust version (which encodes similar IR design concepts) — so how the IR is structured is actually very flexible. The things I’ve been playing with have sort of been linear — but the structure of the IR in general is like a tree.

I could easily imagine thinking about an optimization pass which ignores non-operation based loop constructs (like implicit loops defined by branching between blocks). One question I did have was about the data available during saturation — is it possible to represent equivalences which are type dependent? E.g. I may write a generic “addition” operation which is further specialized after a type inference pass. It seems like the main thing with including type information is that it could blow up the number of rules, or otherwise get quite complicated if there was a subtyping relationship, etc. Probably that concern is completely separate from the discussion here, and again comes back to encoding.

One thing that this discussion sparked was possibly encoding E-PEG using the IR concepts, and more generally I would be curious to understand the relationship of loop analysis in E-PEG compared to polyhedral techniques.

So it seems like two takeaways — 
1. ignoring implicit loops, can use rewrite transformations on straight like code. 
2. maybe explore encoding E-PEG using the design concepts of the IR

**Reply by chandrakananandi** [reply]
Yes I have heard about MLIR although I am not very familiar with the details. @gussmith23 @slyubomirsky may know more about that space.

I wonder if you can encode type information not in the rule itself but perhaps as an analysis which you can then use to apply the rules using something like a conditional applier. If that is possible, you will not have to create too many specializations of the same generic rule.

**Reply by femtomc** [reply]
That’s a great idea! Thanks Chandra!

**Reply by Bastacyclop** [reply]
Regarding types, an alternative to using e-class analyses is to embed the types in the e-graph (the easiest with egg would probably be to include types in terms/e-nodes, in my own project I actually include types in e-classes). Then, it is possible to write untyped (or partially annotated) rewrite rules and to implement rewrite rule type inference: producing a typed rule from the untyped one. A typed rule can match (deconstruct) types on its left-hand-side, and construct types on its right-hand-side. It is also possible for rewrite rule type inference to generate appliers on the right-hand-side if purely syntactic constructs are not sufficient. I am successfully using this approach for a typed lambda calculus language.

**Reply by michel-steuwer** [reply]
You can read about @Bastacyclop's work in our paper that you find on arXiv: https://arxiv.org/pdf/2111.13040.pdf

We are very interested in bringing equality saturation to MLIR and we have a couple of contacts at Google working with MLIR who are very interested in this as well.

**Reply by femtomc** [reply]
@michel-steuwer Cool! My project is not really affiliated with MLIR in any formal way -- I'm just inspired by their IR design, and I'm exploring some PL questions associated with the design concepts. The way I'm experimenting with things now, I actually split types out of the IR structure -- and I'm putting them into an abstract interpretation framework. I think that means that it will be harder in my case to get the right information to the right places (at least, it requires a bit more thought).

**Reply by femtomc** [reply]
(on the other hand, the reason why I reached out to `egg` is because my experiment is written in Rust, so it seemed easier to integrate now, versus whatever is required to provide the same capabilities to C++)
